### DISCUSSION TITLE
Avoid XXE vulnerabilities in Java SAX XML parser

### DISCUSSION THREAD

**Discussion by kennethwong-hc** [crawler-commons/crawler-commons#370] [Q&A] [answered]
Hi team,

I saw in https://github.com/crawler-commons/crawler-commons/commit/7555bcbbbec0c44731896d5b9ac9ef4451b375d2 that below 2 features were set to false to avoid XXE vulnerabilities:
```
            factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
            factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);

```
but why don't just set http://apache.org/xml/features/disallow-doctype-decl to true? as [this ](https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html#java)mentioned this is the safest method.

Thanks.

**Comment by sebastian-nagel** [comment, marked as answer]
Hi @kennethwong-hc, I agree, the simplest solution would be the best. However, DTDs may appear in RSS feeds (the [RSS RFC v1.0](https://validator.w3.org/feed/docs/rss1.html) includes an example) which are allowed as [alternative sitemap formats](https://www.sitemaps.org/protocol.html#otherformats). If there is an DTD in a feed or sitemap and disallow-doctype-decl is true, parsing will fail, see also the comments in #323.

Is there any missing/overseen security feature in the SAX parser factory configuration?

We might decide to disallow DTDs at all but would need to communicate/document that appropriately.

**Reply by kennethwong-hc** [reply]
Hi @sebastian-nagel , thanks for your reply and the clarification of not adding this feature.
How about adding this feature if a particular property value has been set?
Rometools has similar thing done.
https://github.com/rometools/rome/commit/44a17c3456d9503f4071d3bd3c58921a57df608b

**Reply by sebastian-nagel** [reply]
Yes, of course, good idea. - Despite that it makes the code even more complex. We might even think about dropping support for DTDs at all - but needs a discussion.

Thanks, for the link to the Rometools issue. Feel free to open an issue for this. If you have an actual proof-of-concept at hand even better.

**Reply by sebastian-nagel** [reply]
Sorry, I've missed #371. - If you want to add the option there, also fine.

**Reply by kennethwong-hc** [reply]
Change has been made on https://github.com/crawler-commons/crawler-commons/pull/371
